crates/sui-graphql-rpc/src/types/transaction_block/tx_lookups.rs

```diff
@@ -0,0 +1,468 @@
+    /// adjusted to the smaller of the two. The resulting value is additionally modified by the
+    /// `scan_limit` if `is_from_front` is true.
+    pub(crate) fn scan_hi(&self) -> u64 {
+        let adjusted_hi = self.tx_hi();
+
+        if self.is_from_front {
+            if let Some(scan_limit) = self.scan_limit {
+                adjusted_hi.min(self.tx_lo().saturating_add(scan_limit))
+            } else {
+                adjusted_hi
+            }
+        } else {
+            adjusted_hi
+        }
+    }
+
+    /// If the query result does not have a previous page, check whether the scan limit is within
+    /// the initial tx_sequence_number range.
+    pub(crate) fn scan_has_prev_page(&self) -> bool {
+        if self.after.unwrap_or(0) >= self.hi {
+            return false;
+        }
+
+        self.scan_lo() > self.lo
+    }
+
+    /// If the query result does not have a next page, check whether the scan limit is within the
+    /// initial tx_sequence_number range.
+    pub(crate) fn scan_has_next_page(&self) -> bool {
+        if self.before.unwrap_or(self.hi) <= self.lo {
+            return false;
+        }
+
+        self.scan_hi() < self.hi
+    }
+
+    pub(crate) fn scan_prev_page_and_lo(&self) -> (bool, u64) {
+        let has_prev = self.scan_has_prev_page();
+
+        (has_prev, self.scan_lo())
+    }
+
+    pub(crate) fn scan_next_page_and_hi(&self) -> (bool, u64) {
+        let has_next = self.scan_has_next_page();
+
+        (has_next, self.scan_hi())
+    }
+}
+
+impl TransactionBlockFilter {
+    /// Most filter conditions require a scan limit if used in tandem with other filters. The
+    /// exception to this is sender and checkpoint, since sender is denormalized on all tables, and
+    /// the corresponding tx range can be determined for a checkpoint.
+    pub(crate) fn requires_scan_limit(&self) -> bool {
+        [
+            self.function.is_some(),
+            self.kind.is_some(),
+            self.recv_address.is_some(),
+            self.input_object.is_some(),
+            self.changed_object.is_some(),
+            self.transaction_ids.is_some(),
+        ]
+        .into_iter()
+        .filter(|is_set| *is_set)
+        .count()
+            > 1
+    }
+
+    /// With the exception of `transaction_ids`, other filter conditions are denormalized with the
+    /// sender info, and are handled there.
+    pub(crate) fn requires_explicit_sender(&self) -> bool {
+        self.transaction_ids.is_some()
+            || (self.function.is_none()
+                && self.kind.is_none()
+                && self.recv_address.is_none()
+                && self.input_object.is_none()
+                && self.changed_object.is_none())
+    }
+
+    /// A TransactionBlockFilter is considered not to have any filters if no filters are specified,
+    /// or if the only filters are on `checkpoint`.
+    pub(crate) fn has_filters(&self) -> bool {
```

Comment:
This function is similar to the `requires_scan_limit` one, but you've written it a different way -- I think both ways are fine, but we should write both functions the same way -- your pick!

Reply:
while they are similar, they check different things. `has_filters` checks if we need to query lookup tables, while `requires_scan_limit` checks whether we have at least 2 complex filters (which would require the user to provide a scan limit)